src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfoOfT.cs

```diff
@@ -106,10 +110,117 @@ internal sealed class JsonPropertyInfo<T> : JsonPropertyInfo
             PropertyTypeCanBeNull = DeclaredPropertyType.CanBeNull();
             _propertyTypeEqualsTypeToConvert = typeof(T) == DeclaredPropertyType;
 
-            GetPolicies(ignoreCondition, parentTypeNumberHandling, defaultValueIsNull: PropertyTypeCanBeNull);
+            GetPolicies(ignoreCondition, parentTypeNumberHandling);
+        }
+
+        internal void InitializeForSourceGen(
+            JsonSerializerOptions options,
+            bool isProperty,
+            Type declaringType,
+            JsonTypeInfo typeInfo,
+            JsonConverter converter,
+            Func<object, T> getter,
+            Action<object, T> setter,
+            JsonIgnoreCondition ignoreCondition,
+            JsonNumberHandling numberHandling,
+            string propertyName,
+            JsonEncodedText jsonPropertyName)
```

Comment:
It the idea we can cache common property names together on the caller site, instead of here on the callee side which wouldn't know about the duplicated names?

Reply:
Yes that's the idea. The source generator/caller will keep track of duplicate property names and resolve them within the callback to return the list of `JsonPropertyInfo`s. We still detect collisions when putting the infos in the dictionary for deserialization, and throw `IOE` for those cases.

However, it will be possible for the same property name to be written multiple times if the caller is only serializing (and not deserializing). The source generator will not create metadata that has this behavior, but the user can construct this themselves unless we decide to validate both sides (essentially populating the deserialization cache even though we don't need to)